### PR TITLE
OpenGL sprite batch drawing

### PR DESCRIPTION
**Pull request by Overv**
This is an effort to change the OpenGL renderer such that it combines draw operations into batches that are executed with a single draw call.

I've changed all of the `OpenGLDrawContext` operations to add the parameters of their draw operations to a queue instead of executing them directly. This allows the renderer to potentially optimize draw calls by combining them.

Note: Peeps switching colors already happened before the new array based texture cache

### CODE REVIEW

**Comment by LRFLEW** [comment]
`glDebugMessageCallback` was added in OpenGL 4.3. We're primarily targeting OpenGL 3.2 right now, and we shouldn't be using functions introduced later than that. Also, the highest version of OpenGL available in macOS right now is 4.1, so that function wouldn't be available anyways. Maybe conditionally support it (as in not making the whole renderer fail if it's not available) would work, but it would be better to check for the extension string to do that.

**Comment by LRFLEW** [comment]
Be careful with GLSL. Since every driver has a slightly different compiler, relying on certain behaviors can be problematic. For example, my driver (macOS Nvidia GPU) doesn't like [drawimage.frag:30](https://github.com/OpenRCT2/OpenRCT2/pull/4147/files#diff-1be7e8b3e69b6f5da9b8a66592b71b42R30), requiring the expression to be explicitly a bool. Changing it to `if (fMask != 0)` fixes the issue.

**Comment by IntelOrca** [comment]
> For example, my driver (macOS Nvidia GPU) doesn't like drawimage.frag:30, requiring the expression to be explicitly a bool. Changing it to if (fMask != 0) fixes the issue.

Same.

**Comment by mrtnptrs** [comment]
I know this is work in progress, but when I switch to the OpenGL engine, I get these errors:

```
ERROR[src\drawing\NewDrawing.cpp:99 (drawing_engine_init)]: Unable to initialise drawing engine. Falling back to software.
Error compiling C:\Users\maart\AppData\Local\Temp\Rar$EXa0.249\data\shaders\drawimage.frag
0(30) : error C7011: implicit cast from "int" to "bool"

ERROR[src\drawing\NewDrawing.cpp:99 (drawing_engine_init)]: Unable to initialise drawing engine. Falling back to software.
```

My GPU supports OpenGL 4.5, so, that isn't the problem here.

**Comment by Overv** [comment]
@mrtnptrs That is caused by the issue @LRFLEW and @IntelOrca mentioned. I will fix it ASAP.

**Comment by Overv** [comment]
@LRFLEW I can fix the `glDebugMessageCallback` and `glTexStorage3D` calls, but I can not get rid of `glVertexAttribDivisor`. I have tried to multiply the data, but it very negatively affects performance.

**Comment by danidoedel** [comment]
Just had a crash while testing this, not sure if it's related or needed but I leave it here anyway (I maximized the window and the game crashed):
[12efa111-acb9-4663-97fb-bb828c82a20d(9185c4c).zip](https://github.com/OpenRCT2/OpenRCT2/files/379797/12efa111-acb9-4663-97fb-bb828c82a20d.9185c4c.zip)

**Comment by Overv** [comment]
@danidoedel That is normal, the array implementation of the texture cache was limited to 2048 textures, so it broke after a few intro parks.

**Comment by janisozaur** [comment]
@Overv Travis is failing, can you please see to that?

```
/home/travis/build/OpenRCT2/OpenRCT2/src/drawing/engines/opengl/OpenGLDrawingEngine.cpp: In member function ‘virtual void OpenGLDrawingEngine::Initialise(SDL_Window*)’:
/home/travis/build/OpenRCT2/OpenRCT2/src/drawing/engines/opengl/OpenGLDrawingEngine.cpp:297:9: error: ‘PFNGLDEBUGMESSAGECALLBACKPROC’ was not declared in this scope
         PFNGLDEBUGMESSAGECALLBACKPROC debugMessageCallback = (PFNGLDEBUGMESSAGECALLBACKPROC) SDL_GL_GetProcAddress("glDebugMessageCallback");
         ^
/home/travis/build/OpenRCT2/OpenRCT2/src/drawing/engines/opengl/OpenGLDrawingEngine.cpp:297:39: error: expected ‘;’ before ‘debugMessageCallback’
         PFNGLDEBUGMESSAGECALLBACKPROC debugMessageCallback = (PFNGLDEBUGMESSAGECALLBACKPROC) SDL_GL_GetProcAddress("glDebugMessageCallback");
                                       ^
/home/travis/build/OpenRCT2/OpenRCT2/src/drawing/engines/opengl/OpenGLDrawingEngine.cpp:298:13: error: ‘debugMessageCallback’ was not declared in this scope
         if (debugMessageCallback != nullptr) {
             ^
make[2]: *** [CMakeFiles/openrct2.dir/src/drawing/engines/opengl/OpenGLDrawingEngine.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/openrct2.dir/all] Error 2
make: *** [all] Error 2
```

File: src/rct2.c [hunk 1]
```diff
@@ -301,21 +301,20 @@ void rct2_draw(rct_drawpixelinfo *dpi)
...
-		_currentFPS = 1000.0f / frameDelta;
-	}
-	_lastFPSTicks = currentTicks;
-	return _currentFPS;
+    if (time(NULL) != _lastSecond) {
+        _currentFPS = _frames;
+        _frames = 0;
+    }
+
+    _lastSecond = time(NULL);
```
Comment (IntelOrca) [inline src/rct2.c]: No need to call time(NULL) again... stash the last result in a variable.

Comment (Overv) [inline src/rct2.c]: Good point, also fixes a possible race condition

**Comment by janisozaur** [comment]
@Overv travis is still failing:
```
In file included from /home/travis/build/OpenRCT2/OpenRCT2/src/drawing/engines/opengl/TextureCache.cpp:22:0:
/home/travis/build/OpenRCT2/OpenRCT2/src/drawing/engines/opengl/TextureCache.h:165:5: error: array must be initialized with a brace-enclosed initializer
     };
     ^
```

```
/home/travis/build/OpenRCT2/OpenRCT2/src/drawing/engines/opengl/OpenGLDrawingEngine.cpp:297:55: error: 
      unknown type name 'GLDEBUGPROC'
        typedef void (APIENTRYP debugMessageCallback)(GLDEBUGPROC callba...
                                                      ^
/home/travis/build/OpenRCT2/OpenRCT2/src/drawing/engines/opengl/OpenGLDrawingEngine.cpp:300:36: error: 
      cannot initialize a parameter of type 'int' with an lvalue of type 'void
      (GLenum, GLenum, GLuint, GLenum, GLsizei, const GLchar *, const void *)'
            glDebugMessageCallback(OpenGLAPI::DebugCallback, nullptr);
                                   ^~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
make[2]: *** [CMakeFiles/openrct2.dir/src/drawing/engines/opengl/OpenGLDrawingEngine.cpp.o] Error 1
make[1]: *** [CMakeFiles/openrct2.dir/all] Error 2
```

For whatever reason, it does compile on my machines. The log on the one with GeForce GPU looks like so: https://gist.github.com/janisozaur/2b71594ed55dd122215e1b5be505844c.

Eventually the space in atlases depletes and this happens:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  no atlas with free slots left that supports image!
```

**Comment by Overv** [comment]
I will fix the Travis errors as soon as the functionality is finished. I've almost finished dynamic allocation of more atlases.

**Comment by Overv** [comment]
Dynamic atlas allocation is finished, I'm going to fix the build errors now.

**Comment by Overv** [comment]
I don't know what else to optimise in terms of OpenGL code right now, so this is ready to be checked and merged. Can people test to see if it works well?

There are still texture, transparency and UI glitches in the OpenGL renderer, but these also exist on the develop branch. They were not caused by the new rendering code and I currently don't know how to fix them.

**Comment by janisozaur** [comment]
`SetUncappedFrameRate` needs an `override`

**Comment by janisozaur** [comment]
Tested on Linux with GeForce/proprietary drivers and Intel/mesa (GL 4.3): :+1:

**Comment by LRFLEW** [comment]
I made Overv/OpenRCT2#2 to fix the issues with the Xcode project. I've tested this, and once that patched is added, then I give this my Ok. :+1:

**Comment by janisozaur** [comment]
@Overv can you please amend the merge commit and remove `#2` from it or just completely forego it and simply merge the other one? If not done, this will pollute our issue tracker with needless issue-mentions once merged.

**Comment by Overv** [comment]
@LRFLEW Yes, I only took the existing implementation and made it faster. I didn't look into any of the issues I noticed yet.
@janisozaur Fixed.

**Comment by mrtnptrs** [comment]
https://github.com/OpenRCT2/OpenRCT2/wiki/OpenGL-renderengine-issues

**Comment by Overv** [comment]
@mrtnptrs I think nearly 50% of those issues are caused by the same problem of not properly handling colour palettes, which we discussed in the chat yesterday.

File: src/windows/options.c [hunk 2]
```diff
@@ -587,6 +587,7 @@ static void window_options_mouseup(rct_window *w, int widgetIndex)
 		switch (widgetIndex) {
 		case WIDX_UNCAP_FPS_CHECKBOX:
 			gConfigGeneral.uncap_fps ^= 1;
+            drawing_engine_set_fps_uncapped(gConfigGeneral.uncap_fps);
```
Comment (duncanspumpkin) [inline src/windows/options.c]: as this is a c file it needs tabs instead of spaces.

**Comment by duncanspumpkin** [comment]
@Overv can you write a brief overview of how this new version works. Maybe write a new wiki entry about it.

**Comment by Overv** [comment]
@duncanspumpkin Sure, I'll try to do that later today.

File: src/rct2.c [hunk 3]
```diff
@@ -301,21 +301,22 @@ void rct2_draw(rct_drawpixelinfo *dpi)
...
+static int _currentFPS;
+static int _frames;
 
-static float rct2_measure_fps()
+static void rct2_measure_fps()
 {
-	uint32 currentTicks = SDL_GetTicks();
-	if (currentTicks - _lastFPSUpdateTicks > 500) {
-		_lastFPSUpdateTicks = currentTicks;
+    _frames++;
```
Comment (duncanspumpkin) [inline src/rct2.c]: Indentation for this file should be tabs as well. (I know its a pain how we have the two styles but its not changing any time soon)

Comment (Overv) [inline src/rct2.c]: Fixed that. Is it not possible to just plan some point in time where the majority of contributors agree to convert all files at once?

**Comment by Overv** [comment]
@duncanspumpkin I have written an article for the wiki:

https://github.com/OpenRCT2/OpenRCT2/wiki/Fast-OpenGL-renderer

**Comment by duncanspumpkin** [comment]
Thanks.

File: src/drawing/engines/opengl/OpenGLDrawingEngine.cpp [hunk 4]
```diff
@@ -794,11 +879,86 @@ void OpenGLDrawingContext::DrawGlyph(uint32 image, sint32 x, sint32 y, uint8 * p
...
+
+    command.bounds[0] = left;
+    command.bounds[1] = top;
+    command.bounds[2] = right;
+    command.bounds[3] = bottom;
+
+    _commandBuffers.images.push_back(command);
+}
+
+void OpenGLDrawingContext::FlushCommandBuffers() {
```
Comment (IntelOrca) [inline src/drawing/engines/opengl/OpenGLDrawingEngine.cpp]: Move brace to its own line.

File: src/drawing/engines/opengl/OpenGLDrawingEngine.cpp [hunk 5]
```diff
@@ -794,11 +879,86 @@ void OpenGLDrawingContext::DrawGlyph(uint32 image, sint32 x, sint32 y, uint8 * p
...
+    FlushRectangles();
+    FlushLines();
+
+    FlushImages();
+}
+
+void OpenGLDrawingContext::FlushRectangles() {
```
Comment (IntelOrca) [inline src/drawing/engines/opengl/OpenGLDrawingEngine.cpp]: Move brace to its own line.

File: src/drawing/engines/opengl/OpenGLDrawingEngine.cpp [hunk 6]
```diff
@@ -794,11 +879,86 @@ void OpenGLDrawingContext::DrawGlyph(uint32 image, sint32 x, sint32 y, uint8 * p
...
+    for (const auto& command : _commandBuffers.rectangles) {
```
Comment (IntelOrca) [inline src/drawing/engines/opengl/OpenGLDrawingEngine.cpp]: Move brace to its own line.

File: src/drawing/engines/opengl/TextureCache.cpp [hunk 7]
```diff
@@ -135,6 +168,38 @@ void * TextureCache::GetImageAsARGB(uint32 image, uint32 tertiaryColour, uint32
...
+    // If there is no such atlas, then create a new one
+    if ((int) _atlases.size() >= _atlasesTextureIndicesLimit) {
+        throw std::runtime_error("more texture atlases required, but device limit reached!");
+    }
+
+    int atlasIndex = (int) _atlases.size();
+    int atlasSize = (int) powf(2, (float) Atlas::CalculateImageSizeOrder(imageWidth, imageHeight));
+
+#ifdef DEBUG
+    printf("new texture atlas #%d (size %d) allocated\n", atlasIndex, atlasSize);
```
Comment (IntelOrca) [inline src/drawing/engines/opengl/TextureCache.cpp]: Use Console::WriteLine or `log_verbose`.

File: src/drawing/engines/opengl/TextureCache.h [hunk 8]
```diff
@@ -17,9 +17,12 @@
 #pragma once
 
 #include <unordered_map>
+#include <vector>
+#include <algorithm>
```
Comment (IntelOrca) [inline src/drawing/engines/opengl/TextureCache.h]: Order standard headers alphabetically.

File: src/drawing/engines/opengl/TextureCache.h [hunk 9]
```diff
@@ -50,24 +53,148 @@ struct GlyphId
...
+// This is the maximum width and height of each atlas, basically the
+// granularity at which new atlases are allocated (2048 -> 4 MB of VRAM)
+constexpr int TEXTURE_CACHE_MAX_ATLAS_SIZE = 2048;
+
+// Pixel dimensions of smallest supported slots in texture atlases
+// Must be a power of 2!
+constexpr int TEXTURE_CACHE_SMALLEST_SLOT = 32;
+
+// Location of an image (texture atlas index, slot and normalized coordinates)
+struct CachedTextureInfo {
```
Comment (IntelOrca) [inline src/drawing/engines/opengl/TextureCache.h]: Put brace on separate line.

File: src/drawing/engines/opengl/TextureCache.cpp [hunk 10]
```diff
@@ -77,46 +76,80 @@ GLuint TextureCache::GetOrLoadGlyphTexture(uint32 image, uint8 * palette)
...
+    }
+}
+
+void TextureCache::EnlargeAtlasesTexture(GLuint newEntries) {
+    CreateAtlasesTexture();
+
+    GLuint newIndices = _atlasesTextureIndices + newEntries;
+
+    // Retrieve current array data
+    std::vector<char> oldPixels(_atlasesTextureDimensions * _atlasesTextureDimensions * _atlasesTextureIndices);
```
Comment (IntelOrca) [inline src/drawing/engines/opengl/TextureCache.cpp]: Preferred syntax: `auto oldPixels = std::vector<char>(_atlasesTextureDimensions * _atlasesTextureDimensions * _atlasesTextureIndices);`

**Comment by IntelOrca** [comment]
Other than style comments, the code looks merge-able, I will test this out tonight.

File: src/drawing/engines/opengl/TextureCache.cpp (hunk 10 above)
Comment (Overv) [inline src/drawing/engines/opengl/TextureCache.cpp]: All of the style issues should be fixed now.

**Pull request merged**